### PR TITLE
Builtin functions refactoring

### PR DESCRIPTION
**Pull request by FLHerne**
Yay, more random cleanups.

Slightly RFC on the decorator bit; if this style makes sense there are probably more things to apply it to. Personally I think anything is better than these big tables.

### CODE REVIEW

File: nml/actions/action2var_variables.py [hunk 1]
```diff
@@ -562,8 +562,14 @@ def industry_town_count(name, args, pos, info):
...
-    from nml.expression.functioncall import builtin_cargotype
-    return (builtin_cargotype(name, args, pos), [])
+    if len(args) != 1:
+        raise generic.ScriptError(name + "() must have 1 parameter", pos)
+    if not isinstance(args[0], expression.StringLiteral) or args[0].value not in global_constants.cargo_numbers:
+        raise generic.ScriptError(
+            "Parameter for {}() must be a string literal that is also in your cargo table", pos
+        )
+    num = expression.ConstantNumeric(global_constants.cargo_numbers[args[0].value])
+    return (num, [])
```
Comment (FLHerne) [inline nml/actions/action2var_variables.py:572]: Ok, so this is *increasing* duplication...I'll try to find a better way to fix this.

(also, I just realized I forgot to format the format-string here...)

Comment (glx22) [reply nml/actions/action2var_variables.py:572]: Can't import `function_table` from `functioncall` and use `function_table["cargotype"](name, args, pos)` ?

Comment (FLHerne) [reply nml/actions/action2var_variables.py]: I couldn't then, because `name` wouldn't have been in the lookup table. Sorted now though.

**Review by LordAro — COMMENTED** [review]
Not a huge fan of the "random" distinction between `@builtin` and `@builtin_names`, and with the naming of the functions. Could they be made more consistent? 

I'd probably go with always having a "builtin_" prefix to the function name and always using `@builtin_names`

File: nml/actions/action2var_variables.py [hunk 2]
```diff
@@ -562,8 +562,7 @@ def industry_town_count(name, args, pos, info):
     return (args[0], extra_params)
 
 def industry_cargotype(name, args, pos, info):
-    from nml.expression.functioncall import builtin_cargotype
-    return (builtin_cargotype(name, args, pos), [])
+    return (expression.functioncall.builtin_resolve_typelabel(name, args, pos), [])
```
Comment (frosch123) [inline nml/actions/action2var_variables.py:565]: builtin_resolve_typelabel would not need that "default to cargo" trick, if there was a 4th optional paramater for the function name, like
`
builtin_resolve_typelabel("cargo", args, pos, username=name)
`

Comment (FLHerne) [reply nml/actions/action2var_variables.py:565]: > Not a huge fan of the "random" distinction between `@builtin` and `@builtin_names`, and with the naming of the functions. Could they be made more consistent?
> 
> I'd probably go with always having a "builtin_" prefix to the function name and always using `@builtin_names`

I changed `@builtin` to require and strip off the `builtin_` prefix, which removes most of the inconsistency.

Possibly I'm overthinking this and should just use `builtins("name")` for everything, but this way reduces the chance of copy-paste errors if people add new functions.

**Comment by FLHerne** [comment]
> Not a huge fan of the "random" distinction between `@builtin` and `@builtin_names`, and with the naming of the functions. Could they be made more consistent?
> 
> I'd probably go with always having a "builtin_" prefix to the function name and always using `@builtin_names`

I changed `@builtin` to require and strip off the "builtin_" prefix, which removes most of the inconsistency. Perhaps I'm overthinking this, but it should reduce the chance of copy-paste errors if someone adds a new function.

File: nml/actions/action2var_variables.py (hunk 2 above)
Comment (FLHerne) [reply nml/actions/action2var_variables.py:565]: I really don't like this idea. The existing `name` parameter already has that meaning; passing in a fake value to manipulate the behaviour and then adding a new parameter for the real value seems like a far bigger hack. The default is going to exist somewhere, why not here?

(I updated the comment a bit)

**Review by frosch123 — APPROVED** [review]

**Pull request merged**